### PR TITLE
Extract lib transactions

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #1139

A facade for database transactions, and injection of an implementation at startup time, breaks some dependencies from ProjectHistory, UndoManager, AudioIO.

This pull requests depends on #1429

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"

### CODE REVIEW

**Comment by crsib** [comment]
I'm not sure that I understand the need for this library. While it makes total sense to me to make the persistence layer abstract and, at least, to move it out of the main source tree, I feel that TransactionScope as a separate library with dependency injection is just an unnecessary complication.

**Comment by Paul-Licameli** [comment]
> I'm not sure that I understand the need for this library. While it makes total sense to me to make the persistence layer abstract and, at least, to move it out of the main source tree, I feel that TransactionScope as a separate library with dependency injection is just an unnecessary complication.

I found this extraction desirable as I as extracting libraries.  I want to have a library for project file IO, and one for AudioIO, the playback and recording engine.  I don't want the second one depending on the first one.  This small facade breaks that dependency.

**Comment by Paul-Licameli** [comment]
> > I'm not sure that I understand the need for this library. While it makes total sense to me to make the persistence layer abstract and, at least, to move it out of the main source tree, I feel that TransactionScope as a separate library with dependency injection is just an unnecessary complication.
> 
> I found this extraction desirable as I as extracting libraries. I want to have a library for project file IO, and one for AudioIO, the playback and recording engine. I don't want the second one depending on the first one. This small facade breaks that dependency.

I may describe other justifications for this if I stare at the right graphs again of intermediate states of my big branch.

**Comment by Paul-Licameli** [comment]
Yes, another explanation is that in the current arrangement, UndoManager as well as others use DBConnection.cpp and avoid dependency on ProjectFileIO.  There is not a dependency cycle.

But to factor the libraries without this facade, putting ProjectFileIO and DBConnection together -- which seems the right thing to do -- would require that all of ProjectHistory and UndoManager would be sucked into that library too.  That, however, feels wrong.  I would like ProjectHistory with UndoManager in their own library.  The small Transactions facade, rather than all of DBConnection, is the lower level thing it uses.  It lets the project history library avoid direct or transitive dependency on sqlite.

**Comment by crsib** [comment]
Now it feels even more wrong...

* Why do we abstract out only the transaction then? Changing the persistence layer is still not possible with this approach and transaction simply is a part of persistence.
* Why AutoSave is a part of a transaction? It has nothing to do with transactions in the current code and it is unlikely that it will ever be. I don't agree that it should be there at all.
* UndoHistory is a direct user of the database, which probably should be fixed. But it is a user of the library, rather a part of it.

File: libraries/lib-transactions/TransactionScope.cpp
Comment (crsib) [inline libraries/lib-transactions/TransactionScope.cpp:71]: I have a strong opinion, that implementation should throw or otherwise provide the user with details on what to do, rather than a generic message that something has failed. The user side of error handling is already way too confusing and unhelpful for the user

File: libraries/lib-transactions/TransactionScope.cpp
Comment (crsib) [inline libraries/lib-transactions/TransactionScope.cpp:83]: This feels like a possible source of broken projects, although rollback failures are rather rare

**Comment by Paul-Licameli** [comment]
> > I'm not sure that I understand the need for this library. While it makes total sense to me to make the persistence layer abstract and, at least, to move it out of the main source tree, I feel that TransactionScope as a separate library with dependency injection is just an unnecessary complication.
> 
> I found this extraction desirable as I as extracting libraries. I want to have a library for project file IO, and one for AudioIO, the playback and recording engine. I don't want the second one depending on the first one. This small facade breaks that dependency.

But you don't like the dependency injection of a global object providing the services?

We already have this technique with BasicUI.  It may be understandable and justifiable there.  Yet I can understand unease about overusing it.

In my self-review I also realized I don't like mentioning AudacityProject as an incomplete type, necessary as an argument of the function, because Project belongs to another library not yet extracted.

So what if instead:  this little library describes a TransactionServices abstract class that doesn't mention the project.  UndoManager is constructed with one of those things, and the global AudioIO object is given one of them when you start a stream.  The object can make an abstract TransactionScope and also do the AutoSave.

There is still some need, as I explained, to lower some abstract class into this library, to break the dependencies.

**Comment by crsib** [comment]
>So what if instead: this little library describes a TransactionServices abstract class that doesn't mention the project

Sounds good to me!

**Comment by Paul-Licameli** [comment]
> Now it feels even more wrong...
> 
> * Why do we abstract out only the transaction then? Changing the persistence layer is still not possible with this approach and transaction simply is a part of persistence.
> * Why AutoSave is a part of a transaction? It has nothing to do with transactions in the current code and it is unlikely that it will ever be. I don't agree that it should be there at all.
> * UndoHistory is a direct user of the database, which probably should be fixed. But it is a user of the library, rather a part of it.

Yes, there are things in the existing design that feel wrong.  TransactionScope was added in places for certain performance reasons, because everywhere else we do small auto-commits by default.  But as we have mentioned before, we really need to reorganize the sqlite usage to make properly coarse grained transactions.

This is another load of work to do, but I don't want to make the other progress in extraction of libraries block on it.  Fixing that part of design is not the scope of this.

Ideally the framework would handle the beginning and ending of transactions better, and TransactionScope won't be needed for peformace and would disappear.

But the third commit in this sequence will still have done something worth preserving.  In the better arrangement, the Undo manager will need a way to call back to the database and tell it to commit.  And as I explained, putting the Undo manager into a separate library not dependent on the details of the persistence layer makes a lot of sense to me.

File: libraries/lib-transactions/TransactionScope.cpp
Comment (Paul-Licameli) [reply libraries/lib-transactions/TransactionScope.cpp:71]: A fair point but a separate question from the proposal in this PR.  This much of exsitent behavior is preserved without change, though moved into another file.

File: libraries/lib-transactions/TransactionScope.cpp
Comment (Paul-Licameli) [reply libraries/lib-transactions/TransactionScope.cpp:83]: Rollback as the RAII step in a desctructor makes sense.  A destructor should also not throw.  It should do no-fail operations.  Yet we can't quite be sure rollback is always no-fail?  (A decent database should try very hard to make it so!)

No totally acceptable thing to do for this rare eventuality.  But again, it's beside the purpose of this PR, which does not make this problem better or worse than it was.

File: libraries/lib-transactions/TransactionScope.cpp
Comment (crsib) [reply libraries/lib-transactions/TransactionScope.cpp:71]: I understand, that this somewhat maintains the current behavior. My point is - such separation will make it hard to fix it in the future.

File: libraries/lib-transactions/TransactionScope.cpp
Comment (crsib) [reply libraries/lib-transactions/TransactionScope.cpp:83]: >A destructor should also not throw

This is true 99.9% of the time, but probably this one is in 0.1%? :-) C++ really fails us here.

I think we should rewrite transactions in the future (?) to accept the function object, that should be run as a part of the transaction. Like `io->Transaction([this](UndoStuff();});`

**Comment by Paul-Licameli** [comment]
> > So what if instead: this little library describes a TransactionServices abstract class that doesn't mention the project
> 
> Sounds good to me!

But I don't know if I have yet convinced you that it is good for the Undo manager to call into this layer at all?

I explained that the pushing of Undo history state should be the place where database transactions commit, as the AutoSave function.  I don't think that will change.  I think that in the properly redesigned system, this will also be the place that the next transaction is opened.

And rollbacks?  There exists a ProjectHistory::RollbackState function which has been called that (by me) since even before sqlite was used, because an analogy with database rollback seemed natural.  See how AudacityApp invokes it from a global exception handler for the event loop.  See how it's also done in many cases where you were doing a shift-drag of something in the Track panel, but quit with the Esc key.

I think in the ideal design, that RollbackState would also be the place for actual rollback of sqlite transactions.  There will be a larger "transaction" encompassing both in-memory project state and database state.  There are places where we want to guarantee their consistency with commits or rollbacks.  Undo history will be the place.

So in short:  while the scoped transactions may go away in future, I think these points in the life of the un-scoped transactions will still make a lot of sense in Undo manager, and it makes sense to give it access to that layer through some interface.

**Comment by crsib** [comment]
> But I don't know if I have yet convinced you that it is good for the Undo manager to call into this layer at all?

Oh, I have no problems with that for sure. What I meant is that the fact UndoManager calls into the DB layer doesn't make it a part of the DB layer. Probably I'm missing the dependency cycle though

**Comment by Paul-Licameli** [comment]
> > But I don't know if I have yet convinced you that it is good for the Undo manager to call into this layer at all?
> 
> Oh, I have no problems with that for sure. What I meant is that the fact UndoManager calls into the DB layer doesn't make it a part of the DB layer. Probably I'm missing the dependency cycle though

Project history depends on transactions Interface as client.  File io depends on it as supplier.  That is clear.

It would be desirable to make this separation, just so that project history depends on less, but maybe you think it is not needed to break a cycle.

But file io will depend on project history too, although the reason for that is very slender.

See how SqliteSampleBlock.cpp uses Undo manager, just so that it can make a progress dialog during purge of undo history, using events emitted by Undo manager.

**Comment by Paul-Licameli** [comment]
> > > But I don't know if I have yet convinced you that it is good for the Undo manager to call into this layer at all?
> > 
> > 
> > Oh, I have no problems with that for sure. What I meant is that the fact UndoManager calls into the DB layer doesn't make it a part of the DB layer. Probably I'm missing the dependency cycle though
> 
> Project history depends on transactions Interface as client. File io depends on it as supplier. That is clear.
> 
> It would be desirable to make this separation, just so that project history depends on less, but maybe you think it is not needed to break a cycle.
> 
> But file io will depend on project history too, although the reason for that is very slender.
> 
> See how SqliteSampleBlock.cpp uses Undo manager, just so that it can make a progress dialog during purge of undo history, using events emitted by Undo manager.

You know what?  Listening to myself, and looking at other commits in my branch, I'm doubting my own story.  I worked hard to achieve a dependency inversion between project history and the persistence layer, really because I was trying to make project history not depend on subclasses of Track (Wave track in particular).  But maybe that's all unnecessary.

I may need to rethink other things in the branch.  Platform neutral project file IO might need fewer steps to achieve after all.

However, I don't want AudioIO dependent on the persistence layer.  The existing use of a TransactionScope inside it is a reason why it would.  So, that smaller reason still remains, why this library is useful, until such time as we do the other work to avoid the need for that TransationScope.

**Comment by Paul-Licameli** [comment]
Pinging @crsib to review this branch again.

It is now a smaller PR.  The intention is only to allow later extraction of AudioIO in a library, which I hope can happen in this release.  I'm no longer trying to break dependency of undo and redo on the persistency layer.

Still unaddressed, as mentioned in comments above:  the uninformative message "Database error.  Sorry, but we don't have more details." and the problem that failures detected in the destructor are not better communicated.  But those are existing problems, and this just moves the code.

**Comment by crsib** [comment]
I think we should wait for 3.1.3 to be merged into master (or we can do it now!)

3.1.3 adds a new transaction internally, so we won't be able to merge it cleanly into the master if we merge this PR first

**Comment by Paul-Licameli** [comment]
> I think we should wait for 3.1.3 to be merged into master (or we can do it now!)
> 
> 3.1.3 adds a new transaction internally, so we won't be able to merge it cleanly into the master if we merge this PR first

I don't understand the reservations.  I tried rebasing this branch onto release on my own computer, and there were no serious conflicts.  It should not need to wait either before or after merging 3.1.3 into master.  Where was a transaction added using TransactionScope?

Do you have any objections to the content of the pull request?

**Comment by crsib** [comment]
Sorry, that was part of 3.1.2. Let's merge it then!

**Review by crsib — APPROVED** [review]

**Pull request merged**
